### ISSUE TITLE
VHD support

### ISSUE THREAD

**Issue by Torinde** [dosbox-staging/dosbox-staging#3307]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.81.0-RC

### What Operating System are you using?

Windows 10

### Is your feature request related to a problem? Please describe.

VHD is nice if you want to support Win9x/ME due to the dynamic size support, and it is also better supported on Windows hosts for mounting HDD image files, compared to the raw image files that Staging supports. With VHD files on Windows 8+ hosts you can just right click them and mount them. With RAW HDD files, I only know of 7zip that can open them, and even then only in read-only mode.

_Originally posted by @rderooy in https://github.com/dosbox-staging/dosbox-staging/issues/313#issuecomment-1880003367_

[emulator requirements for eXoWin9x](https://www.retrorgb.com/interview-with-exo-pc-ms-dos-and-windows-game-preservationist.html): "master Win98 install, with each game installing to its own child image. This would allow any changes to the OS folders (or the C: at all) to be captured by the child\diff image." (#2185)

### Describe the solution you'd like

VHD support 
* [ ] Fixed
* [ ] Dynamic
* [ ] Differencing 
* [ ] mounting
* [ ] creating
* [ ] merging differencing VHD(s) into the parent VHD
* [ ] converting from IMG to VHD
* [ ] converting from fixed VHD to dynamic VHD
* [ ] converting from dynamic VHD to fixed VHD
* [ ] timestamp checking option
* [ ] information showing about a VHD image
* [ ] snapshotting UI (maybe as part of OSD)
* [ ] snapshotting capability to that while running (#313)
* [ ] VHDX ([Bochs supports it](https://github.com/bochs-emu/Bochs/pull/717)) - to support metadata journaling, the [LBA 33-bit enhancement for DOS](https://www.bttr-software.de/forum/forum_entry.php?id=17268) that would allow 2x2TB MBR partitions, GPT in DOS or Win9x ([[1]](https://www.bttr-software.de/forum/forum_entry.php?id=19502), [[2]](https://github.com/Baron-von-Riedesel/DOS71BSG), [[3]](https://www.bttr-software.de/forum/forum_entry.php?id=20642), [[4]](https://github.com/LordOfMice/Tools) [GPTTSD](https://msfn.org/board/topic/184646-terabyte-pack-2/))

### Add any other context or screenshots about the feature request here.

VHD differencing and dynamic support was [added to DOSbox-X](https://github.com/joncampbell123/dosbox-x/pull/4273)
- Snapshot-like [arrangement](https://github.com/joncampbell123/dosbox-x/issues/5292) for three VHD layers for Windows/Game/User branching-off
- snapshots tree parent/child info (paths, filenames, hashes, relationship/position in the tree, etc.) can be put in a text file inside the image, so that if you mount a child, DOSBox can automatically load the relevant parent files by getting their details from this file.
- would be nice to combine with https://github.com/dosbox-staging/dosbox-staging/issues/313 for RAM/CPU/emulator

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
Thanks @Torinde . That's a bit of a laundry list of a feature request, but I guess there's no harm in asking ;) I imagine external programs could do all that conversion business, no need to overload the DOSBox codebase with all that stuff.

**Comment by rderooy** [comment]
[qemu-img](https://www.qemu.org/docs/master/tools/qemu-img.html) can be used to convert RAW HDD images to VHD.

DOSBox-X has a 'imgmake' program to create raw FDD and HDD images, and VHD HDD images. That functionality would be useful to have in Staging to create image files.

**Comment by shermp** [comment]
I've looked into this in the past. It requires a bit more extensive refactoring than "just" adding VHD support.

In the interim, I've created myself a little python program that can append and remove a VHD footer to a disk image. I might publish it someday, but it does little error checking, so I'm a bit nervous about sharing it in case it ruins someone's disk image.

**Comment by rderooy** [comment]
> > DOSBox-X has a `imgmake` program to create raw FDD and HDD images, and VHD HDD images. That functionality would be useful to have in Staging to create image files.
> 
> IMGMAKE also can directly create partitioned and formatted images, I think.

That is correct, and is the default behaviour. Formatted FAT12 FDD and partitioned and formatted FAT16 or FAT32 HDD images.

**Comment by emxd** [comment]
> I've looked into this in the past. It requires a bit more extensive refactoring than "just" adding VHD support.
> 
> In the interim, I've created myself a little python program that can append and remove a VHD footer to a disk image. I might publish it someday, but it does little error checking, so I'm a bit nervous about sharing it in case it ruins someone's disk image.

There's also gulikoza's raw2vhd utlility (https://www.vogons.org/viewtopic.php?t=34642#p298235). That + imgmake I use a lot between dosbox/pcem/qemu. I had issues in the past with Windows qemu-img, mounting the images in winexplorer - Windows would refuse to mount them. I think it was related to the sparse representation (https://gitlab.com/qemu-project/qemu/-/issues/136)
